src/api-umbrella/utils/interval_lock.lua

```diff
@@ -0,0 +1,78 @@
+local lock = require "resty.lock"
+
+local _M = {}
+
+--- Only one thread can execute fn through its execution duration
+-- A lock (mutex) is held while the function executes and released at the end
+-- @param name - a unique identifier for this lock (automatically namespaced)
+-- @param fn - function to execute if the lock can be held
+-- @return ok, error
+_M.mutex_exec = function(name, fn)
+  local check_lock = lock:new("locks", {["timeout"] = 0})
+  local _, lock_err = check_lock:lock("mutex_exec:" .. name)
+  if lock_err then
+    return true, nil
+  else
+    local results = pcall(fn)
+    local ok, unlock_err = check_lock:unlock()
+    if not ok then
+      ngx.log(ngx.ERR, "failed to unlock: ", unlock_err)
+    end
+    return results
+  end
+end
+
+--- Only allow a function to be executed once in a given interval
+-- A lock (mutex) is set to expire after an interval of time. If the mutex is
+-- present, execution won't take place
+-- @param name - a unique identifier for this lock (automatically namespaced)
+-- @param interval - the length of time until expiry (seconds)
+-- @param fn - function to execute within the interval
+-- @return ok, error
+_M.timeout_exec = function(name, interval, fn)
+  local ok, err = ngx.shared.interval_locks:add(name, true, interval)
+  if not ok and err == "exists" then
+    return true, nil
+  elseif not ok then
+    return false, err
```

Comment:
Can we add an `ngx.log(ngx.ERR)` statement inside here before returning false? This is a case that really shouldn't happen, but if it does, it likely means something something bad has happened like we've run out space on the `interval_locks` shared memory table. So I think logging an error would be helpful so we at least know if we've hit this unexpected situation.